src/screencast/fps_limit.c

```diff
@@ -0,0 +1,67 @@
+#include "fps_limit.h"
+#include "logger.h"
+#include "timespec_util.h"
+#include <stdint.h>
+#include <time.h>
+#include <unistd.h>
+
+#define FPS_MEASURE_PERIOD_SEC 5.0
+
+void measure_fps(struct fps_limit_state* state, struct timespec *now);
+
+void fps_limit_measure_start(struct fps_limit_state* state, double max_fps) {
+	if (max_fps <= 0.0) 
+		return;
+
+	clock_gettime(CLOCK_MONOTONIC, &state->last_time);
+}
+
+uint64_t fps_limit_measure_end(struct fps_limit_state* state, double max_fps) {
+	if (max_fps <= 0.0) 
+		return 0;
+
+	if (timespec_is_zero(&state->last_time)) {
+		// `fps_limit_measure_start` was not called?
```

Comment:
I am not sure. Do we want to abort the program in case the assertion fails? I mean, limiting the number of frames captured by the application is not a critical feature; in case something goes wrong here, I could imagine that we'd want to continue rather than to crash.

Let me know what you think; if you want me, I can certainly add an assertion here.

Reply:
In general I prefer to fail hard when I end up in an unexpected state, this makes the users report this as an issue and allows us to fix potential bugs instead of ending up with more subtle bugs that are harder to debug.

I'm not feeling strongly about this particular case though, feel free to ignore my comment.

Reply:
I'll do some more thorough testing with the assertion in place after https://github.com/emersion/xdg-desktop-portal-wlr/issues/91 is fixed, to check whether I actually have this condition met in the code. I'll probably keep the condition afterwards.

Reply:
I replaced the check with an assert statement, see https://github.com/emersion/xdg-desktop-portal-wlr/pull/74/commits/f06bc743f00736b8cbd5d25701e1671f62dba000 . I did some further testing, with various screensharing sessions going on and off, and didn't find any issue. I am marking this comment as resolved, please consider reopening if you have further comments.